core/src/main/java/com/netflix/conductor/core/execution/DeciderService.java

```diff
@@ -610,10 +610,11 @@ private Task getDynamicTasks(WorkflowDef def, Workflow workflow, WorkflowTask ta
 		List<String> joinOnTaskRefs = new LinkedList<>();
 		// Create Dynamic tasks
 		for (WorkflowTask wft : dynForkTasks) {
+			joinOnTaskRefs.add(wft.getTaskReferenceName());
```

Comment:
Used the `wft` to add the taskReferenceName instead of using `forkedTasks` as it seems that in some race conditions, when other thread schedules the tasks, the `forkedTasks` is empty for current thread and results into empty `joinOnTaskRefs` which overwrites the join task.